### PR TITLE
perf: optimize building circuit polynomial

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR optimizes building a circuit polynomial.

The main optimization parts are as follows:

- Avoid memory allocations if possible.
   - Change the value type to the reference type.
   - Use `OPENMP_PARALLEL_FOR` instead of `base::Parallelize()`
   - Use `BatchInverseInPlaceSerial()` within an existing parallel loop instead of `BatchInverseInPlace()`.
- Optimize algorithms
   - Use `MulBy2ExpInPlace()` when computing `a * 2ᵉˣᵖ`
   - Avoid calling `ToBigInt()` which incurs montgomery reduction
   - Use `GetCoset()`
   - Use a new feature of `GetSuccesivePowers()` in #221.

### CODE REVIEW

File: tachyon/zk/plonk/vanishing/vanishing_utils.h [hunk 1]
```diff
@@ -21,11 +21,12 @@ namespace tachyon::zk::plonk {
...
 template <typename F>
 constexpr F GetZeta() {
-  CHECK_EQ(F::Config::kTrace % math::BigInt<F::kLimbNums>(3),
-           math::BigInt<F::kLimbNums>(0));
-  return F::FromMontgomery(F::Config::kSubgroupGenerator)
-      .Pow(F(2).Pow(F::Config::kTwoAdicity).ToBigInt() * F::Config::kTrace /
-           math::BigInt<F::kLimbNums>(3));
+  using BigInt = typename F::BigIntTy;
+  CHECK_EQ(F::Config::kTrace % BigInt(3), BigInt(0));
+  BigInt exp = F::Config::kTrace;
```
Comment (dongchangYoo) [inline tachyon/zk/plonk/vanishing/vanishing_utils.h:26]: How about adding a comment like the following
// The result of exponential does not exceed modular of scalar field.

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

**Comment by ashjeong** [comment]
1. https://github.com/kroma-network/tachyon/pull/288/commits/540e0734bd6f0063e4f2e62978a64ef2fd5946bf

**Comment by TomTaehoonKim** [comment]
@chokobole Could you add me as a co-author for [4b63f1a](https://github.com/kroma-network/tachyon/pull/288/commits/4b63f1a89436e4bf5b83327cae6af03935db4ecf) and [b015be3](https://github.com/kroma-network/tachyon/pull/288/commits/b015be3eb6b5bfa4acf3059c8339cb1ac16f2f7b) please?

File: tachyon/zk/plonk/vanishing/vanishing_utils.h [hunk 2]
```diff
@@ -190,7 +190,7 @@ std::vector<Evals> CoeffsToExtendedPart(const Domain* domain,
 
 template <typename F>
 std::vector<F> BuildExtendedColumnWithColumns(
-    std::vector<std::vector<F>>&& columns) {
+    const std::vector<std::vector<F>>& columns) {
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/zk/plonk/vanishing/vanishing_utils.h:170]: This is because there's no where to use this moved `columns`

**Comment by chokobole** [comment]
> @chokobole Could you add me as a co-author for [f04850a](https://github.com/kroma-network/tachyon/pull/288/commits/f04850afea3c1ef553bde8242f1a90782922ccfd) and [b015be3](https://github.com/kroma-network/tachyon/pull/288/commits/b015be3eb6b5bfa4acf3059c8339cb1ac16f2f7b) please?

I'll change the author to you!

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Pull request merged**